### ISSUE TITLE
feat: docs-json: always resolve types

### ISSUE THREAD

**Issue by kraftwer1** [ionic-team/stencil#4583]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already include this feature request, without success.

### Describe the Feature Request

As of today, docs-json can resolve types (like unions). Example:

```
@Prop() background: Background
```

...results in:

```
background: "dark" | "white“
```

However, it does not resolve them when used in an event's detail:

```
@Event() backgroundChange: EventEmitter<{ background: Background }>
```

...results in:

```
backgroundChange: { detail: { background: Background } }

// Expected: { detail: { background: "dark" | "white" } }
```

It also does not resolve when used as an array:

```
@Prop() backgrounds: Background[]
```

...results in:

```
@Prop() backgrounds: Background[]

// Expected: backgrounds: ("dark" | "white")[]
```

Furthermore, it seems that more complex structures aren't resolved, either:

```
@Prop() config: { background: Background }
```

...results in:

```
config: { background: Background }
// Expected: config: { background: "dark" | "white" }
```

### Describe the Use Case

It would help the consumers to read the docs without having to see our TypeScript types. Most of the consumers use plain JS without TypeScript and therefore the types are more confusing than helping. Even worse: the types even "hide" the important information from them. They'd never know that `Background` is actually `"dark" | "white"`.

### Describe Preferred Solution

Long story short: always resolve types. Or see "Alternative" below.

### Describe Alternatives

Instead of always resolve types, docs-json could also append some kind of a "Types database" and then it'd be okay to have the types on the docs. Example:

```
// Generated json

{
  "timestamp": "...",
  "compiler": "...",
  "components": [{
    "tag": "my-background",
    "props": [{
      "name": "background",
      "type": "Background"
    }]
  }],
  "types": [{ // <-- this
    "id": "Background",
    "value": "\"dark\" | \"white\""
  }]
}
```

**Comment by rwaskiewicz** [comment]
Relates to https://github.com/ionic-team/stencil/issues/4394, https://github.com/ionic-team/stencil/issues/3603

**Comment by alicewriteswrongs** [comment]
@kraftwer1 thanks for filing this issue - would you mind adding some info about which version of Stencil you're using? I'm just asking because in Stencil 4 we added a feature to expand the amount of information included in the `docs-json` output target (see here: https://github.com/ionic-team/stencil/pull/4212). Are you already on Stencil 4 by chance? If so, it would be great if you could put together a reproduction case showing how you're using `@Event` and not getting satisfactory output.

**Comment by kraftwer1** [comment]
@alicewriteswrongs you're right. I was using 3.2.2. After upgrading to 4.0.1, I can see you've added `"typeLibrary"`. I think I can work with that for now, however I'll have to do all the lookup and fiddling myself. Is Stencil planning to do that in the future?

**Comment by alicewriteswrongs** [comment]
For now we don't have plans to do more than providing the information, but hopefully that's enough to at least get a richer documentation experience together!

Probably you already noticed, but the types referenced in `@Method`, `@Prop`, `@Event` and so on properties are referenced via an id that you can then look up in the type library.

I'm going to go ahead and close this issue now, good luck using these documentation features to build out your project!
